### PR TITLE
*: prepare and recover flashback env

### LINKED ISSUE #37376: prepare & recover env

### PR DESCRIPTION
**Pull request by Defined2014**
### What problem does this PR solve?

Issue Number: close #37376

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: domain/infosync/schedule_manager.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+// GetPDScheduleConfig get schedule config from schedules map
+func (mm *mockScheduleManager) GetPDScheduleConfig(ctx context.Context) (map[string]interface{}, error) {
+	mm.Lock()
+	defer mm.Unlock()
+
+	return mm.schedules, nil
+}
+
+// SetPDScheduleConfig set schedule config to schedules map
+func (mm *mockScheduleManager) SetPDScheduleConfig(ctx context.Context, config map[string]interface{}) error {
```
Comment (bb7133) [inline domain/infosync/schedule_manager.go:84]:
```suggestion
func (mm *mockScheduleManager) SetPDScheduleConfig(ctx context.Context, config map[string]interface{}) {
```
If there's no way to return an error, we could remove `error` from the signature?

File: domain/infosync/schedule_manager.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	mm.Lock()
+	defer mm.Unlock()
```
Comment (bb7133) [inline domain/infosync/schedule_manager.go:86]: Please avoid using `defer` when it is not necessary.

File: domain/infosync/schedule_manager.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+	return nil
+}
+
+type mockScheduleManager struct {
+	sync.RWMutex
+	schedules map[string]interface{}
+}
+
+// GetPDScheduleConfig get schedule config from schedules map
+func (mm *mockScheduleManager) GetPDScheduleConfig(ctx context.Context) (map[string]interface{}, error) {
```
Comment (bb7133) [inline domain/infosync/schedule_manager.go:76]:
```suggestion
func (mm *mockScheduleManager) GetPDScheduleConfig(ctx context.Context) (map[string]interface{}) {
```

File: meta/meta.go [hunk 4]
```diff
@@ -59,25 +59,26 @@ var (
...
+	mRandomIDPrefix     = "TARID"
+	mBootstrapKey       = []byte("BootstrapKey")
+	mSchemaDiffPrefix   = "Diff"
+	mPolicies           = []byte("Policies")
+	mPolicyPrefix       = "Policy"
+	mPolicyGlobalID     = []byte("PolicyGlobalID")
+	mPolicyMagicByte    = CurrentMagicByteVer
+	mDDLTableVersion    = []byte("DDLTableVersion")
+	mConcurrentDDL      = []byte("concurrentDDL")
+	mInFlashbackCluster = []byte("InFlashbackCluster")
```
Comment (bb7133) [inline meta/meta.go:81]: IMHO it is not an elegant design to introduce `mInFlashbackCluster`, but it is straightforward and easy for quick-and-safe implementation.

I can accept it for now.

File: domain/infosync/schedule_manager.go (hunk 3 above)
Comment (Defined2014) [reply domain/infosync/schedule_manager.go:76]: The return type follow interface `ScheduleManager`

File: domain/infosync/schedule_manager.go (hunk 1 above)
Comment (Defined2014) [reply domain/infosync/schedule_manager.go:84]: ditto

File: domain/infosync/schedule_manager.go (hunk 2 above)
Comment (Defined2014) [reply domain/infosync/schedule_manager.go:86]: We should lock the map to avoid concurrency issues. Although it will never happen in Flashback Cluster.

**Review by bb7133 — APPROVED** [review]
LGTM

**Review by xhebox — COMMENTED** [review]
Overall LGTM. But `flashback` may needs to wait for tikv or PD. It sounds like it needs a middle state to wait.

**Pull request merged**